libsolidity/interface/FileReader.cpp

```diff
@@ -0,0 +1,124 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <libsolidity/interface/FileReader.h>
+
+#include <liblangutil/Exceptions.h>
+
+#include <libsolutil/Exceptions.h>
+#include <libsolutil/CommonIO.h>
+#include <libsolutil/CommonData.h>
+
+using solidity::frontend::ReadCallback;
+using solidity::langutil::InternalCompilerError;
+using solidity::util::errinfo_comment;
+using solidity::util::readFileAsString;
+using std::optional;
+using std::string;
+using std::vector;
+
+namespace solidity {
+
+vector<string> FileReader::sourceNames() const
+{
+	vector<string> names;
+	for (auto const& source: m_sourceCodes)
+		names.emplace_back(source.first);
+	return names;
+}
+
+void FileReader::setBasePath(boost::filesystem::path const& _path)
+{
+	m_basePath = _path;
+	allowDirectory(_path);
+}
+
+void FileReader::allowDirectory(boost::filesystem::path _path)
+{
+	if (!util::contains(m_allowedDirectories, _path))
+		m_allowedDirectories.emplace_back(std::move(_path));
+}
+
+void FileReader::setSource(string _path, optional<boost::filesystem::path> _fspath, string _source)
```

Comment:
The `_path` here is not necessarily a path in general case. I think we should have a specific term for it because it gets confusing quickly. In #11105 I proposed calling it "source unit ID" (actually I used "import key" at first but then changed it). Another option could be `VFS key`. What's your opinion on that?

```suggestion
void FileReader::setSource(string _sourceUnitID, optional<boost::filesystem::path> _filesystemPath, string _sourceCode)
```